(#37) Honor Godog's life cycle hooks (Before/After Scenario)

mdelapenya opened the pull request:
## What does this PR do?
This PR moves the Before/After Scenario hooks to each test file (Redis, Apache, MySQL) so that each own controls the life cycle of the containers they start.

We are adding the ability to each Service to define a CleanUp function, that will be triggered after a successful termination of the container.

On the other hand, as a refactoring task, we are moving assertions to its own Go file, so that they can become an assertions library to be reused across tests in a simpler manner, cleaning up the test runner.

## Why is it important?
We want each test to be idempotent, which means starting the containers they need Before each scenario, and destroying them After each scenario. That why we can ensure we do not put dirty data into the storage, so that each test creates and destroys the data it produces.


## Related Issues
Closes #37

jsoriano commented on metricbeat-tests/mysql_test.go:
Same code is being added to all feature contexts, I think this should be part of some helper.

jsoriano commented on metricbeat-tests/mysql_test.go:
Same here, this is added for all services, this should probably be part of some helper.

jsoriano commented on metricbeat-tests/mysql_test.go:
```diff
@@ -27,9 +39,14 @@ func metricbeatIsInstalledAndConfiguredForMySQLModule(metricbeatVersion string)
+	err := serviceManager.Run(mysqlService)
+	if err == nil {
+		mysqlTestSuite.Service = mysqlService
+	}
```
And also here, same code added for all services.

jsoriano commented on metricbeat-tests/metricbeat.go:
```diff
@@ -47,10 +50,16 @@ func RunMetricbeatService(version string, monitoredService services.Service) (se
+	fn := func(ctx context.Context) error {
+		return deleteIndex(ctx, "metricbeat", indexName)
+	}
+	service.SetCleanUp(fn)
```
I think that it would be even more important to delete the indexes before running the scenario to ensure that they run in a clean environment.

mdelapenya commented:
Hey @jsoriano I addresses your suggestions in a refactor of the test cases:
- removed implementations for apache, redis and mysql, moving functions to the structure for the test suite.
- keep the initialisation of godog's context and steps
- refactored regex in steps so that it captures service type too, making them reusable

jsoriano commented on metricbeat-tests/apache_test.go:
Great to see this generalized. If we need to add some specific step for a service, I guess that we can still create another context with its own steps? Or we'd need to add the steps to the general `FeatureContext`.

mdelapenya replied:
This is something on my radar, so we must investigate it
